pytorch_lightning/metrics/metric.py

```diff
@@ -255,3 +255,72 @@ def __setstate__(self, state):
         self.__dict__.update(state)
         self.update = self._wrap_update(self.update)
         self.compute = self._wrap_compute(self.compute)
+
+
+class MetricCollection(nn.ModuleDict):
+    """
+    MetricCollection class can be used to chain metrics that have the same
+    call pattern into one single class.
+
+    Args:
+        metrics: one of the following
+            * list or tuple: if metrics are passed in as a list, will use the metrics
+                class name as key for output dict. Therefore, two metrics of the
+                same class cannot be chained this way.
+            * dict: if metrics are passed in as a dict, will use each key in the dict
+                as key for output dict. Use this format if you want to chain together
+                multiple of the same metric with different parameters.
+    Example:
+
+        >>> from pytorch_lightning.metrics import MetricCollection, Accuracy, Precision, Recall
+        >>> target = torch.tensor([0, 2, 0, 2, 0, 1, 0, 2])
+        >>> preds = torch.tensor([2, 1, 2, 0, 1, 2, 2, 2])
+        >>> metrics = MetricCollection([Accuracy(),
+        ...                             Precision(num_classes=3, average='macro'),
+        ...                             Recall(num_classes=3, average='macro')])
+        >>> metrics(preds, target)
+        {'Accuracy': tensor(0.1250), 'Precision': tensor(0.0667), 'Recall': tensor(0.1111)}
+
+        >>> metrics = MetricCollection({'micro_recall': Recall(num_classes=3, average='micro'),
+        ...                             'weighted_recall': Recall(num_classes=3, average='macro')})
+        >>> metrics(preds, target)
+        {'micro_recall': tensor(0.1250), 'weighted_recall': tensor(0.1111)}
+
+    """
+    def __init__(self, metrics):
+        super().__init__()
+        if isinstance(metrics, dict):
+            # Check all values are metrics
+            for name, metric in metrics.items():
+                if not isinstance(metric, Metric):
+                    raise ValueError(f'Value {metric} belonging to key {name}'
+                                     ' is not an instance of `pl.metrics.Metric`')
+                self[name] = metric
+        elif isinstance(metrics, (tuple, list)):
+            for metric in metrics:
+                if not isinstance(metric, Metric):
+                    raise ValueError(f'Input {metric} to `MetricCollection` is not a instance'
+                                     ' of `pl.metrics.Metric`')
+                name = metric.__class__.__name__
+                if name in self:
+                    raise ValueError(f'Encountered two metrics both named {name}')
+                self[name] = metric
+        else:
+            raise ValueError('Unknown input to MetricCollection.')
+
+    def forward(self, *args, **kwargs):
```

Comment:
let's say I have two metrics like this:
`Metric1` takes inputs `a` and  `b`. `Metric2` takes inputs `c` and `d`. With your implementations both metrics would get `a`, `b`, `c` and `d`, correct? 

I think the cleaner way would be to pass all args and extract kwargs based on the metric's method signature using inspect.

What do you think?